crates/data-plane-controller/src/repo.rs

```diff
@@ -0,0 +1,147 @@
+use super::{logs, run_cmd};
+use anyhow::Context;
+use std::sync::Arc;
+
+pub struct Repo {
+    inner: Arc<Inner>,
+}
+
+struct Inner {
+    repo: String,
+    idle: std::sync::Mutex<Vec<tempfile::TempDir>>,
+}
+
+pub struct Checkout {
+    dir: Option<tempfile::TempDir>,
+    inner: Arc<Inner>,
+}
+
+impl Repo {
+    pub fn new(repo: &str) -> Self {
+        Self {
+            inner: std::sync::Arc::new(Inner {
+                repo: repo.to_string(),
+                idle: Default::default(),
+            }),
+        }
+    }
+
+    pub async fn checkout(
+        &self,
+        logs_tx: &logs::Tx,
+        logs_token: sqlx::types::Uuid,
+        branch: &str,
+    ) -> anyhow::Result<Checkout> {
+        // Attempt to obtain a pre-created checkout, or clone a new one.
+        let dir = self.inner.idle.lock().unwrap().pop();
+        let dir = if let Some(dir) = dir {
+            () = run_cmd(
+                async_process::Command::new("git")
+                    .arg("clean")
+                    .arg("--force")
+                    .current_dir(dir.path()),
+                "git-clean",
+                logs_tx,
+                logs_token,
+            )
+            .await?;
+
+            dir
+        } else {
+            self.create_clone(logs_tx, logs_token).await?
+        };
+
+        () = run_cmd(
+            async_process::Command::new("git")
+                .arg("fetch")
+                .current_dir(dir.path()),
+            "git-fetch",
+            logs_tx,
+            logs_token,
+        )
+        .await?;
+
+        () = run_cmd(
+            async_process::Command::new("git")
+                .arg("checkout")
+                .arg("--detach")
+                .arg("--quiet")
+                .arg(format!("origin/{branch}"))
+                .current_dir(dir.path()),
+            "git-checkout",
+            logs_tx,
+            logs_token,
+        )
+        .await?;
```

Comment:
I think given that we are detaching the head, a sanity check that we are not working on a dirty directory may be useful, otherwise there is a possibility of conflict. Alternatively we can `git checkout --detach --force` to clean up the dirty state.

The sanity check can be done by running `git status --porcelain` and verifying the output is empty.